[SYCL][Matrix] Move elementwise operation under intel namespace and add joint_matrix_apply.

arnamoy10 opened the pull request:
This patch moves the wi_data and wi_element class (and corresponding
    operations) in the matrix-intel.hpp file, under the
    intel::experimental::matrix namespace.  The original implementation is kept (but soon will be deprecated) to make the existing CUDA test cases work.

dkhaldi commented on sycl/test/check_device_code/matrix/matrix-nvptx-tf32-test.cpp:
```diff
@@ -81,12 +81,16 @@ int main() {
           for (auto i = 0; i < 4; ++i)
-            get_wi_data(sg, sub_b)[i] =
-                round_to_tf32(get_wi_data(sg, sub_b)[i]);
+            sycl::ext::intel::experimental::matrix::get_wi_data(sg, sub_b)[i] =
```
this is a cuda backend test so it should not use intel specific APIs, leave it unchanged until Jack adds the CUDA specific changes.

arnamoy10 replied:
Done

dkhaldi commented on sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:
```diff
@@ -43,73 +43,24 @@ struct joint_matrix {
-    return (jm.cuda_impl.wi_marray[i]);
-#else
-    return wi_element<T, Rows, Cols, Use, Layout, Group>(jm, i);
+#else // NVPTX
```
@JackAKirk, what is the plan on adding CUDA specific implementation? Can we have a CUDA PR that adds implementation for this so we can merge them around the same time?

JackAKirk replied:
I've suggested the cuda implementation in a comment above. It is simple and can be added here if you like.

I think the best way forward is 
- Keep the nvidia implementation of wi_data/jm.get_wi_data() but deprecate them (if you want to move wi_data to intel namespace that is fine, nvidia implementation of wi_data can live in that namespace whilst it is deprecated because it is not constructable anyway and this won't break any code). I don't think you have to touch the nvidia wi_data/get_wi_data code at all: you can just move them to the intel header as you've done and intel namespace if you desire this.
- Add the simple nvidia implementation I suggested in this PR for `joint_matrix_apply`.

Then at a later time we can simply remove the deprecated implementations of wi_data and get_wi_data for cuda only.

arnamoy10 replied:
@JackAKirk Please take a look if you are OK with this PR.

JackAKirk replied:
I'll update a cuda test shortly to test everything is fine.

arnamoy10 replied:
Resolving as @JackAKirk confirmed test is working fine.

dkhaldi reviewed (commented):
LGTM other than reverting the nvptx-tf32 test changes.

JackAKirk commented on sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp:
```diff
@@ -290,11 +308,57 @@ class wi_element<sycl::ext::oneapi::bfloat16, NumRows, NumCols, Use, Layout,
+  };
+
+  decltype(auto) operator[](size_t i) {
+    return wi_element<T, Rows, Cols, Use, Layout, Group>(jm, i);
```
So you've moved `wi_data` to the intel header but it is still in the oneapi namespace. I'm not sure it is a good idea to immediately remove the nvidia implementation from wi_data immediately. There is code that is already using this extension and it could be wiser to keep the implementation for now, whilst deprecating it and adding the nvidia implementation to `joint_matrix_apply` at the same time.

Also, if you move wi_data to the intel namespace then you can either just still do the above (existing code use won't care since wi_data is returned as a auto placeholder anyway), and you could even deprecate it in the oneapi namespace and add it to the intel name space also whilst only implementing it there for intel backends.

arnamoy10 replied:
@JackAKirk , Are you OK if I keep the NVIDIA implementation in the oneapi namespace (in `matrix-unified.hpp`) and the `jm` based implementation in the `intel` namespace in `matrix-intel.hpp`?  Thank you

JackAKirk replied:
Yes I think so, although I'm not completely sure I understand what you mean.
Basically from the Nvidia implementation point of view I don't think it matters so long as the changes don't immediately break code that is using `get_wi_data()`. Then wherever the Nvidia implementation is moved to it will be soon removed anyway, once we have given people time to switch to using the `joint_matrix_apply`.

Thanks

JackAKirk replied:
Basically so long as changes don't break this test: https://github.com/intel/llvm-test-suite/blob/intel/SYCL/Matrix/element_wise_all_ops_cuda.cpp

arnamoy10 replied:
@JackAKirk I can keep the NVIDIA/CUDA implementation (I hope they mean the same thing?) in the intel namespace.  In that case, we just have to modify the `get_wi_data()` call to use the `intel` namespace.  Are you OK with that code change in the test case?  I was leaning towards keeping the CUDA implementation in the `oneapi` namespace (in the m`matrix-unified.hpp`) so that we do not have to update the `get_wi_data()` call in the existing NVIDIA test cases.  Let me know what you think.

JackAKirk replied:
Ah I see, yeah you are right I think that `get_wi_data()` for nvidia should stay in the oneapi namespace for Nvidia if that is OK. Point being that if it moved it would break existing code.

Yeah by Nvidia and CUDA backends I mean the same thing. Thanks

JackAKirk commented on sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:
```diff
@@ -43,73 +43,24 @@ struct joint_matrix {
+joint_matrix_apply(Group sg, joint_matrix<Group, T, Use, M, N, Layout> &C,
+                   F &&lambda) {
+#if defined(__SYCL_DEVICE_ONLY__)
 #if defined(__NVPTX__)
```
This is the cuda implementation. Happy for you to add it here.
```suggestion
#if defined(__NVPTX__)
  for (int i = 0; i < jm.cuda_impl.wi_marray.size(); i++) {
    lambda(jm.cuda_impl.wi_marray[i]);
  }
```

dkhaldi commented on sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp:
```diff
@@ -65,14 +65,29 @@ template <typename Group, typename T, use Use, size_t Rows, size_t Cols,
+
+namespace intel::experimental::matrix {
+
+// sycl::ext::oneapi::experimental::matrix::joint_matrix
```
remove this comment

dkhaldi commented on sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:
```diff
@@ -59,21 +59,9 @@ class wi_data {
-    return __spirv_JointMatrixWorkItemLengthINTEL(jm.spvm);
-#endif
-  };
+  size_t length() { return jm.cuda_impl.wi_marray.size(); };
```
I think you need to keep if defined NVPTX
if the else: generate an error that users should use the intel namespace. But not sure whether we can generate static error like this.
@yubingex007-a11y, what do you suggest in this case?

arnamoy10 replied:
Added runtime_error

yubingex007-a11y commented on sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:
```diff
@@ -112,6 +112,29 @@ get_wi_data(Group sg, joint_matrix<Group, T, Use, Rows, Cols, Layout> &jm) {
+#else // NVPTX
+  auto wi_data_c = sycl::ext::intel::experimental::matrix::get_wi_data(sg, jm);
+  for (int i = 0; i < wi_data_c.length(); i++) {
+    lambda(wi_data_c[i]);
```
wi_data's operator[] is returning a temporary object but lambda take a lvalue as input. it will lead to compfail.
the following code can work:
  auto wi_data_c = sycl::ext::intel::experimental::matrix::get_wi_data(sg, jm);
  for (int i = 0; i < wi_data_c.length(); i++) {
    auto wie = wi_data_c[i];
    lambda(wie);
  }

dkhaldi replied:
Besides this, this should be added to update joint matrix values:
"wi_data_c[i] = wie;"
Also, C should be passed by reference to joint_matrix_apply

dkhaldi replied:
This is the full code that runs correctly:
```c++
void joint_matrix_apply(Group sg, joint_matrix<Group, T, Use, M, N, Layout>& C, F&& lambda) {
     auto wi_data_c = get_wi_data(sg, C);
     for (int i = 0; i < wi_data_c.length(); i++) {
       T wie = wi_data_c[i];
       lambda(wie);
       wi_data_c[i] = wie;
     }
     return;
}
```

arnamoy10 replied:
Updated

arnamoy10 commented:
Here is the PR for a test case with these changes: https://github.com/intel/llvm-test-suite/pull/1636

dkhaldi commented on sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp:
```diff
@@ -290,11 +306,57 @@ class wi_element<sycl::ext::oneapi::bfloat16, NumRows, NumCols, Use, Layout,
+          sycl::ext::oneapi::experimental::matrix::use Use, size_t Rows,
+          size_t Cols, sycl::ext::oneapi::experimental::matrix::layout Layout>
+inline __SYCL_ALWAYS_INLINE decltype(auto)
+get_wi_data(Group sg, sycl::ext::oneapi::experimental::matrix::joint_matrix<
```
this also should only work for __NVPTX__ path

dkhaldi commented on sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:
```diff
@@ -63,15 +63,15 @@ class wi_data {
     return jm.cuda_impl.wi_marray.size();
 #else
-    return __spirv_JointMatrixWorkItemLengthINTEL(jm.spvm);
+    throw runtime_error("Use the correct namespace.", PI_ERROR_INVALID_DEVICE);
```
Does it really trigger this?

dkhaldi commented on sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:
```diff
@@ -63,15 +63,15 @@ class wi_data {
     return (jm.cuda_impl.wi_marray[i]);
 #else
-    return wi_element<T, Rows, Cols, Use, Layout, Group>(jm, i);
+    throw runtime_error("Use the correct namespace.", PI_ERROR_INVALID_DEVICE);
```
a better message would be:
"get_wi_data is available using:  ext::intel::experimental::matrix::get_wi_data"

arnamoy10 replied:
Done

dkhaldi commented on sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:
```diff
@@ -112,6 +112,31 @@ get_wi_data(Group sg, joint_matrix<Group, T, Use, Rows, Cols, Layout> &jm) {
 #endif // defined(__SYCL_DEVICE_ONLY__)
 }
 
+template <typename Group, typename T, use Use, size_t M, size_t N,
```
On line 98: an another guard is needed so get_wi_data does not work for JIT path

dkhaldi replied:
Bing suggested to use "(__attribute__((unavailable))" to get compilation error when things are not available on the JIT path.

arnamoy10 replied:
Updated

yubingex007-a11y commented on sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:
```diff
@@ -112,6 +112,31 @@ get_wi_data(Group sg, joint_matrix<Group, T, Use, Rows, Cols, Layout> &jm) {
+template <typename Group, typename T, use Use, size_t M, size_t N,
+          layout Layout, typename F>
+inline __SYCL_ALWAYS_INLINE void
+joint_matrix_apply(Group sg, joint_matrix<Group, T, Use, M, N, Layout> &jm,
```
please handle unused parameter

dkhaldi replied:
unused sg and jm on the host side. Please use same thing like: std::ignore = sg;

JackAKirk reviewed (approved):
CUDA side of things LGTM. Existing tests pass and I have tested joint_matrix_apply on the cuda backend. I have not finished the cuda tests for joint_matrix_apply but I will add a PR for this to intel/llvm-test-suite soon.

JackAKirk commented:
@arnamoy10 Can you just add the standard warning that `get_wi_data` is deprecated on the cuda backend?
Thanks

arnamoy10 commented:
> @arnamoy10 Can you just add the standard warning that `get_wi_data` is deprecated on the cuda backend?
> 
> Cuda backend test for joint_matrix_apply here: [intel/llvm-test-suite#1655](https://github.com/intel/llvm-test-suite/pull/1655)
> 
> Thanks

Thank you for the test case.  I guess we can merge the test case once this patch is merged.

Do you have any example on how to put out the warning?  I checked the file and did not find anything I can use.

JackAKirk commented:
> Do you have any example on how to put out the warning? I checked the file and did not find anything I can use.

I think that this is the normal way of outputing the deprecation: https://github.com/intel/llvm/blob/cf245f623d20aa9c748773e2f7c1e0c6f100deb5/sycl/include/sycl/accessor.hpp#L776

So I thing if you just refer users to joint_matrix_apply in the deprecated message that would be good.

arnamoy10 commented:
> > Do you have any example on how to put out the warning? I checked the file and did not find anything I can use.
> 
> I think that this is the normal way of outputing the deprecation:
> 
> https://github.com/intel/llvm/blob/cf245f623d20aa9c748773e2f7c1e0c6f100deb5/sycl/include/sycl/accessor.hpp#L776
> 
> So I think if you just refer users to joint_matrix_apply in the deprecated message that would be good.

I added the message, can you please double check if I added in the right place?  Thank you

JackAKirk commented:
> > > Do you have any example on how to put out the warning? I checked the file and did not find anything I can use.
> > 
> > 
> > I think that this is the normal way of outputing the deprecation:
> > https://github.com/intel/llvm/blob/cf245f623d20aa9c748773e2f7c1e0c6f100deb5/sycl/include/sycl/accessor.hpp#L776
> > 
> > So I think if you just refer users to joint_matrix_apply in the deprecated message that would be good.
> 
> I added the message, can you please double check if I added in the right place? Thank you

LGTM!

dkhaldi reviewed (approved):
LGTM

steffenlarsen reviewed (approved):
LGTM!

bader commented:
@arnamoy10, please, fix these warnings emitted by `clang` compiler.
https://github.com/intel/llvm/actions/runs/4440014936/jobs/7793257940
```
/Users/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:132:26: error: unused parameter 'sg' [-Werror,-Wunused-parameter]
joint_matrix_apply(Group sg, joint_matrix<Group, T, Use, M, N, Layout> &jm,
                         ^
/Users/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:132:73: error: unused parameter 'jm' [-Werror,-Wunused-parameter]
joint_matrix_apply(Group sg, joint_matrix<Group, T, Use, M, N, Layout> &jm,
                                                                        ^
/Users/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:133:24: error: unused parameter 'lambda' [-Werror,-Wunused-parameter]
                   F &&lambda) {
                       ^
3 errors generated.
```

arnamoy10 commented:
> @arnamoy10, please, fix these warnings emitted by `clang` compiler. https://github.com/intel/llvm/actions/runs/4440014936/jobs/7793257940
> 
> ```
> /Users/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:132:26: error: unused parameter 'sg' [-Werror,-Wunused-parameter]
> joint_matrix_apply(Group sg, joint_matrix<Group, T, Use, M, N, Layout> &jm,
>                          ^
> /Users/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:132:73: error: unused parameter 'jm' [-Werror,-Wunused-parameter]
> joint_matrix_apply(Group sg, joint_matrix<Group, T, Use, M, N, Layout> &jm,
>                                                                         ^
> /Users/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:133:24: error: unused parameter 'lambda' [-Werror,-Wunused-parameter]
>                    F &&lambda) {
>                        ^
> 3 errors generated.
> ```

Thanks for the report.  PR here: https://github.com/intel/llvm/pull/8684
